engines/e_pkcs11.c

```diff
@@ -0,0 +1,841 @@
+
+    tmpl[0].type = CKA_CLASS;
+    tmpl[0].pValue = &key_class;
+    tmpl[0].ulValueLen = sizeof(key_class);
+    tmpl[1].type = CKA_KEY_TYPE;
+    tmpl[1].pValue = &key_type;
+    tmpl[1].ulValueLen = sizeof(key_type);
+    if (ctx->id != NULL) {
+        tmpl[2].type = CKA_ID;
+        tmpl[2].pValue = ctx->id;
+        tmpl[2].ulValueLen = strlen((char *)ctx->id);
+    } else {
+        tmpl[2].type = CKA_LABEL;
+        tmpl[2].pValue = ctx->label;
+        tmpl[2].ulValueLen = strlen((char *)ctx->label);
+    }
+
+    rv = pkcs11_funcs->C_FindObjectsInit(ctx->session, tmpl, OSSL_NELEM(tmpl));
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsInit failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_INIT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjects(ctx->session, &objhandle, 1, &count);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjects failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjectsFinal(ctx->session);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsFinal failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FINAL_FAILED);
+        goto err;
+    }
+
+
+    ctx->key = objhandle;
+    return 1;
+
+ err:
+    return 0;
+}
+
+static char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+static int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2bin;
+    size_t vlen, i, j = 0, tmplen;
+
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    tmplen = strlen(tmp);
+    *value = OPENSSL_malloc(tmplen + 1);
+
+    if (*value == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_PARSE_URI, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    end = strpbrk(tmp, ";");
+
+    BIO_snprintf(*value, end == NULL ? tmplen + 1 :
+             (size_t) (end - tmp + 1), "%s", tmp);
+    hex2bin = OPENSSL_malloc(strlen(*value) + 1);
+
+    if (hex2bin == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_PARSE_URI, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    vlen = strlen(*value);
+    for (i = 0; i < vlen; i++) {
+        if (*(*value+i) == '%' && i < (strlen(*value)-2)) {
+            *(hex2bin+j) = pkcs11_hex_int(*(*value+i+1), *(*value+i+2));
+            i += 2;
+        } else {
+            *(hex2bin+j) = *(*value+i);
+        }
+        j++;
+    }
+    *(hex2bin+j) = '\0';
+    *value = hex2bin;
+    return 1;
+
+ err:
+    return 0;
+}
+
+static int pkcs11_get_console_pin(char **pin)
+{
+#ifndef OPENSSL_NO_UI_CONSOLE
+    int i;
+    char *strbuf = OPENSSL_malloc(512);
```

Comment:
better avoid magic numbers, and their repetition.
Just declare a `const int buflen = 512;` and use it after